### DISCUSSION TITLE
Complete Datatype initialization necessary for subscriber when using Java?

### DISCUSSION THREAD

**Discussion by schubinator239** [OpenDDS/OpenDDS#4565] [Q&A] [answered]
Hello,

I try to subscribe a topic with a complex datatype using a listener in Java.
There first I ran into errors until I discovered the following testcode:
https://github.com/OpenDDS/OpenDDS/blob/master/java/tests/complex_idl/ComplexIDLTest.java

In this testcode the complete datatype is initialized with default values before taking the data with the data holder ( createDefaultData() - line 157 ).

I adapted this to my code and the subscription works now.

Nevertheless, for very complex datatypes this can become a very extensive work.
Is there a way to avoid the default initialization of the subscribed datatype or is there a default method, which already does the job?

I also created a C++ subscriber, where I do not need to initialize the datatype before taking it.
So I guess this is a restirction coming from the JNI bridge or I miss something.

Thanks for your help!

**Comment by mitza-oci** [comment, marked as answer]
Unless there's a case where we can't do so, OpenDDS follows the OMG specs.  The OMG's spec for IDL-to-Java, at the time this code was developed, required each member to be non-null and provided no easy way to do that.

I think the newer specs (IDL4-to-Java) address this issue, so a contribution to implement (parts of) that spec would be welcome.
